src/url.cpp

```diff
@@ -457,4 +457,54 @@ namespace Url
         str.resize(dest);
     }
 
+    Url& Url::deparam(const std::unordered_set<std::string>& blacklist)
+    {
+        remove_params(params_, blacklist, ';');
+        remove_params(query_, blacklist, '&');
+        return *this;
+    }
+
+    void Url::remove_params(std::string& str, const std::unordered_set<std::string>& blacklist, const char separator)
+    {
+
+        std::vector<std::string> pieces;
+        size_t previous = 0;
+        for (size_t index = str.find(separator)
+            ; index != std::string::npos
+            ; previous = index + 1, index = str.find(separator, previous))
+        {
+            std::string piece = str.substr(previous, index - previous);
+            std::string name = piece.substr(0, piece.find('='));
+            std::transform(name.begin(), name.end(), name.begin(), ::tolower);
+
+            if (blacklist.find(name) == blacklist.end())
+            {
+                pieces.push_back(piece);
+            }
+        }
+
+        if (previous < str.length())
+        {
+            std::string piece = str.substr(previous);
+            std::string name = piece.substr(0, piece.find('='));
+            std::transform(name.begin(), name.end(), name.begin(), ::tolower);
```

Comment:
In-person discussion revealed that:

1) we're not going to support (at least for now) case-insensitivity for non-ASCII params
2) we don't exactly know what to do with hostnames -- it seems that https://www.ietf.org/rfc/rfc3454.txt says one thing (and mozscape does it -- https://github.com/seomoz/moz-util-py/blob/master/test/test_convert_url.py#L74-L79), but http://unicode.org/faq/idn.html#11 says another. We'll make a ticket to follow up about that.